drivers:platform:mbed: UART/VCOM platform drivers implementation

mphalke opened the pull request:
Implemented the UART and USBCDC (VirtualCOM) functionality for Mbed platform to read/write data over serial link.
USBCDC (VCOM) is supported/tested fully on SDP-K1 platform.

Signed-off-by: mahphalke <Mahesh.Phalke@analog.com>

buha commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,420 @@
+/******************************************************************************/
+#include <stdio.h>
+#include <mbed.h>
+#include <USBCDC.h>
```
where dos this file come from ? it it a separate package dependency ?

mphalke replied:
It comes from the mbed-os library however it is not included in the standard mbed drivers directory and therefor mbed.h file does not include this file. 
The usb drivers are part of separate directory under drivers and added into the project during project creation if mbed_app.json file explicitly includes it.

![image](https://user-images.githubusercontent.com/62383520/152723995-fd222a79-ebbf-4171-892d-1c4412379da5.png)

buha replied:
I'm just wondering if a user just wants basic UART without caring about UART over USB (USBCDC). The driver file will include the USBCDC.h file anyway which is probably not what the user wants. I don't see how else we could go around this other than #ifdefs

mphalke replied:
That's true and only way seems to be using #if-def which doesn't seem to be feasible as selection b/w UART and VCOM is done from init_param structure at run time.

buha commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,420 @@
+					     desc->extra))->uart_port);
+
+	for (i = 0; i < bytes_number; i++)
+		uart->write(&data[i], 1);
```
Looks like you are using the same ->write() function as in the uart_write() blocking function so this is not actually nonblocking ? Maybe I'm missing something

mphalke replied:
Looks like there is a "writable()" function that should have been called before calling "write()" similar to non-blocking read.
I will correct the code.

buha commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,420 @@
+	if (!uart_desc)
+		return -ENOMEM;
+
+	uart_desc->baud_rate = param->baud_rate;
```
I find it interesting that you don't use the uart_init_param fields, except for this one - baud_rate. The unused ones are:
```
/** UART Device ID */
uint8_t	device_id;
/** UART number of data bits */
enum uart_size		size;
/** UART parity */
enum uart_parity	parity;
/** UART number of stop bits */
enum uart_stop		stop;
```

mphalke replied:
Thanks for pointing this. We were always using default values for these parameters however it make sense to update them in the mbed layer for custom needs. I will updated the drivers accordingly.

mphalke commented:
Also the UnbufferedSerial object needs to be changed to BufferedSerial as buffered serial uses internal UART buffer to store the received data, which prevents flushing the old data if new data arrives immediately. I have made this change which was missing in the first commit.

buha reviewed (approved):
Looks good, my only gripe is about that USBCDC.h, which is debatable.

amiclaus commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,496 @@
+#include "mbed_uart.h"
+
+/* Max size for USB CDC packet during transmit/receive */
+#define USB_CDC_MAX_PACKET_SIZE		(64)
```
nit: are `()` really needed?

mphalke replied:
Don't think they are needed. I'll remove them.

amiclaus commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,496 @@
+#define USB_CDC_MAX_PACKET_SIZE		(64)
+
+/* Max allowed length of USB serial number in characters */
+#define USB_SERIAL_NUM_MAX_LENGTH	(100)
```
ditto

mphalke commented:
V3:

- Removed the brackets around macros

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,496 @@
+{
+	volatile static bool *tx_in_progress = &_tx_in_progress;
+
+	if (*tx_in_progress)
+		return false;
+	else
+		return true;
```
`return !(*tx_in_progress);` might be easier here.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,496 @@
+		usb_cdc_dev = (platform_usbcdc *)((struct mbed_uart_desc *)(
+				desc->extra))->uart_port;
+
+		while (!usb_cdc_dev->data_received(bytes_number)) {
+			/* Wait until new data is available */
+		}
```
Not necessary right now, but it might be good to have a timeout.

mphalke replied:
This is blocking read so not sure if we need timeout? Also it is difficult to estimate timeout value as it depends upon target MCU clock frequency and instruction set, so may be put it as MAX_INT32 if needed at all?

mphalke commented:
V4:

- Modified the return logic of data_transmited() method.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+		usb_cdc_dev = (platform_usbcdc *)((struct mbed_uart_desc *)(
+				desc->extra))->uart_port;
+
+		while (bytes_number) {
```
nitpick: `while (true)` is more correct, if what we want to do is wait for a previous transfer to complete. We never know where bugs might arise.

mphalke commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+		uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+						   desc->extra))->uart_port);
+
+		for (cnt = 0; cnt < bytes_number; cnt++)
+			uart->read(data + cnt, 1);
```
I think that's right, we can use that way. I was facing some issue earlier while reading all bytes simultaneously so used 1 byte read approach. But I will test and confirm this again.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+					   desc->extra))->uart_port);
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart->readable() > 0)
```
If `uart->readable()` returns number of bytes available it should be queried first and if not enough, then exit directly with return 0, in my opinion. The way it is now it will try to read a number of times equal to the number requested, read an arbitrary number of bytes and return that it's done. It should only read and return a a positive, > 0 number if it has actually done so.

mphalke replied:
What readable() mbed implementation says is, it will "return true if there is something to read else return false". It does not return number of bytes. Return type of readable is bool. I think I made a mistake there. I should check for true or false instead of > 0.
It does not explicitly mention or allow to pass number of bytes available to read. It's bit tricky I guess.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
```
Is there a virtual COM option for this one?

mphalke replied:
We haven't figured that out yet Or not implemented. I have mentioned in the function description.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+					   desc->extra))->uart_port);
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart ->writable() > 0)
```
Same as the read: we should query once at the beginning if it's writable, and, if yes, `uart->write(data, bytes_number);`. There may be some checks here that we can actually transfer that many.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+	}
+
+	// Create a new mbed descriptor to store new UART instances
+	mbed_uart_desc = (struct mbed_uart_desc *)calloc(1, sizeof(*mbed_uart_desc));
```
Can we do this before the previous `if` and include the following `if` in the previous one? If it makes sense it will make for a smoother flow, in my opinion.

mphalke replied:
Yes. I will update the code.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+	/* Free the UART object */
+	if (((struct mbed_uart_desc *)desc->extra)->virtual_com_enable) {
+		if ((platform_usbcdc *)((struct mbed_uart_desc *)desc->extra)->uart_port)
+			delete((platform_usbcdc *)(platform_usbcdc *)((struct mbed_uart_desc *)
```
Double cast here.

adrimbarean reviewed (commented):
Some questions and suggestions for now.

mphalke commented:
V5:

- Corrected the readable() and writable() funcion calling logic. Error checks are added for the same.
- Corrected blocking read logic. Calling read(data, cnt) to read all bytes instead of single byte at a time.
- Corrected descriptor/instances memory freeing order in the uart_init() function.
-  Other minor edits as per code review comments.

mphalke commented:
Hi All,
Are we good with the changes or are there any more comments which I need to address?

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+		return 0;
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart->readable())
+			uart->read(&data[i], 1);
```
I'd say we should `break` on the `else` path and return `i` instead of `bytes_number`. Because the danger here is that the user will ask for, let's say 5 bytes, and the buffer has 2 bytes. The function will read 2 bytes and then stay a little while more in the function and return that it has read 5 bytes, which might cause errors.

mphalke replied:
Make sense. I'll update the code.

adrimbarean commented on drivers/platform/mbed/mbed_uart.cpp:
```diff
@@ -0,0 +1,492 @@
+		return -EIO;
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart ->writable())
+			uart->write(&data[i], 1);
```
I'd say things are easier here. We just have to account for the maximum number of bytes that can be queued to be written at once. If `writable`, compare `bytes_number` with the maximum that can be queued at once and queue the smallest number. Then return the number of bytes queued. The for shouldn't be necessary here.

mphalke replied:
Not sure if I get you correctly here. 
Do you mean:
if (uart->writable())
		return uart->write(data, bytes_number);
else
		return 0;

adrimbarean replied:
Yes, this ca work, but does `uart->write` have a maximum number of bytes that can be transmitted at once? From the documentation it seems that it will automatically write a partial write if the requested transfer is too big. So we can do just:

if (uart->writable())
return uart->write(data, bytes_number);

Or it might even work without `if` because if it can't write anything it will return `-EAGAIN`. I'd say both can work.
Note: `uart->write` will return the number of bytes it transmitted, or queued, even if it's just a partial transfer.

mphalke replied:
Okay. I will implement the first one. I was also thinking, we can do same for non-blocking read. If ```readable()``` is true, we will just return ```read()``` with number of bytes else will return 0.
e.g. 
```
if (readable())
    return read(data, bytes)
else
   return 0
```
Let me know if this is okay to implement

adrimbarean replied:
This seems good. Thank you!

adrimbarean reviewed (commented):
I'd say we still need to address two things before we can properly merge here.

mphalke commented:
V6:

- Updated the UART non-blocking read and write functions. ```readable() or writable()``` conditional check is added before ```read/write``` functions. The number of read/write bytes are returned based on the available UART buffer size without blocking.

adrimbarean reviewed (approved):
Looks good to me.
